firedrake/slate/slate.py

```diff
@@ -185,6 +186,15 @@ def __eq__(self, other):
     def __ne__(self, other):
         return not self.__eq__(other)
 
+    @abstractproperty
+    def _complexity(self):
+        """Returns an integer totaling the relevant complexity
+        of the tensor expression. This is done by summing the
+        complexity of each tensor in the expression. This
+        value is a function of operator precedence if the
+        expression is not a terminal tensor.
```

Comment:
Hmmm, this is why I am concerned by this code smell.  Why does operator precedence (say) have anything to do with expression complexity?